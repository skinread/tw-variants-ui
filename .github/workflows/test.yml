# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and run tests across different versions of node
# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs

name: CI Tests

on:
  pull_request:
    branches: ["main","develop"]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [20.x]
    steps:
    - name: Checkout
      uses: actions/checkout@v4
    - name: Setup Node ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        cache-dependency-path: "**/package-lock.json"
    - name: Cache or restore dependencies
      uses: actions/cache@v4
      with:
        path: ~/.npm
        key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
        restore-keys: ${{ runner.os }}-npm-
    - name: Install dependencies
      run: npm ci

    - name: Check formatting and linting
      run: npm run prettier && npm run lint

    - name: Check package build
      run: npm run build

    - name: Get Playwright version
      run: echo "PLAYWRIGHT_VERSION=$(jq -r '.devDependencies["@playwright/test"]' package.json)" >> $GITHUB_ENV
    - name: Cache or restore Playwright binaries
      uses: actions/cache@v4
      with:
        path: |
          ~/.cache/ms-playwright
        key: ${{ runner.os }}-playwright-${{ env.PLAYWRIGHT_VERSION }}
        restore-keys: ${{ runner.os }}-playwright-
    - name: Install Playwright
      if: steps.playwright-cache.outputs.cache-hit != 'true'
      run: npm run pretest
    - name: Storybook test build
      run: npm run storybook:build -- --test
    - name: Storybook test suite
      run: npm run test:ci
