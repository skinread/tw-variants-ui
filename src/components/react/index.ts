export * from './Autocomplete/Autocomplete';
export * from './Button/Button';
export * from './ButtonLink/ButtonLink';
export * from './Checkbox/Checkbox';
export * from './Feedback/Feedback';
export * from './Icon/Icon';
export * from './InputField/InputField';
export * from './Modal/Modal';
export * from './RadioButtons/RadioButtons';
export * from './Select/Select';
export * from './TextArea/TextArea';
/* generator:new component */
