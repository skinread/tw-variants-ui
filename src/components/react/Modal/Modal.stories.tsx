import type { Meta, StoryObj } from '@storybook/react';
import { expect, userEvent, within } from '@storybook/test';

import { Button } from '@components/react';
import { Modal } from './Modal';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  title: 'React/Modal',
  component: Modal,
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
  args: {
    id: undefined,
    title: 'Modal title',
  },
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Modal>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Basic: Story = {
  args: {
    children: <p>Press ESC key or click outside to close</p>,
    id: 'modal-demo-basic',
  },
  decorators: [
    (Story) => (
      <div>
        <Button
          onClick={() => {
            (document.getElementById('modal-demo-basic') as HTMLDialogElement).showModal();
          }}
        >
          Open demo
        </Button>
        <Story />
      </div>
    ),
  ],
};

const triggerText = 'Modal trigger';
export const InteractionTest: Story = {
  args: {
    ...Basic.args,
    id: 'interaction-test',
  },
  decorators: [
    (Story) => (
      <div>
        <Button
          onClick={() => {
            (document.getElementById('interaction-test') as HTMLDialogElement).showModal();
          }}
        >
          {triggerText}
        </Button>
        <Story />
      </div>
    ),
  ],
  play: async ({ args, canvasElement, step }) => {
    const canvas = within(canvasElement);
    const button = canvas.getByText(triggerText);
    const modal = canvas.getByRole('dialog');
    const closeButton = canvas.getByLabelText('close modal');

    async function clickWithDelay(e: Element) {
      await userEvent.click(e);
      await new Promise((resolve) => setTimeout(resolve, 330));
    }

    await step('Initial state', async () => {
      await expect(modal).not.toBeVisible();
    });

    await step('Display modal', async () => {
      await clickWithDelay(button);
      await expect(modal).toBeVisible();
      await expect(modal).toHaveTextContent(args.title!);
      await expect(modal).toHaveTextContent(modal.getElementsByTagName('p')[0]!.textContent!);
      await expect(closeButton).toHaveFocus();
    });

    await step('Close modal', async () => {
      await clickWithDelay(closeButton);
      await expect(modal).not.toBeVisible();
    });
  },
};
