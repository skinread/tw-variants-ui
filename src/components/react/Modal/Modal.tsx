import { ComponentPropsWithRef, forwardRef, type PropsWithChildren } from 'react';
import { Icon } from '@components/react';

const localeDefault = {
  labelClose: 'close modal',
};

type ModalAttributes = PropsWithChildren<ComponentPropsWithRef<'dialog'>>;

export interface ModalProps extends ModalAttributes {
  /**
   * You'll probably need to get the dialog element by id to trigger `showModal()`, unless you're using the shadow DOM. Then use a `ref`.
   */
  id: ModalAttributes['id'];
  /**
   * Optional title text for modal, can be passed through children slot as well
   */
  title?: string;
  /**
   * Content
   */
  children?: ModalAttributes['children'];
  /**
   * Override internal text for localisation, English defaults preset
   */
  localise?: typeof localeDefault;
}

/**
 * The modal component uses the html `<dialog>` element. React `ref` enabled.
 *
 * To trigger the modal it's possible to simply get the element by `id` (or use `ref`) and call the native `showModal()` function.
 * Example usage:
 *
 * ```tsx
 * onClick={() => {
 *  (document.getElementById('modal-id') as HTMLDialogElement)?.showModal();
 * }}
 * ```
 */
export const Modal = forwardRef<HTMLDialogElement, ModalProps>(function VuiModal(props, ref) {
  const { children, id, localise, title, ...attrs } = props;
  const locale = {
    ...localeDefault,
    ...localise,
  };

  return (
    <dialog className="modal items-end sm:items-center" id={id} {...attrs} ref={ref}>
      <div className="modal-box flex flex-col rounded-md p-8 pt-6 w-full max-w-none sm:max-w-xl overflow-hidden">
        <form method="dialog" className="flex justify-end">
          {/* main close button */}
          <button className="btn btn-xs btn-square btn-ghost" aria-label={locale.labelClose}>
            <Icon name="cross" className="stroke-foreground" width={16} />
          </button>
        </form>
        <div className="mt-4 overflow-y-auto">
          {title && <h2 className="font-semibold text-2xl mb-8">{title}</h2>}
          {children}
        </div>
      </div>
      <form method="dialog" className="modal-backdrop" aria-hidden>
        {/* clickable background overlay */}
        <button className="cursor-default" tabIndex={-1}></button>
      </form>
    </dialog>
  );
});
