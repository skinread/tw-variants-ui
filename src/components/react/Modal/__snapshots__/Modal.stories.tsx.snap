// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React/Modal Basic smoke-test 1`] = `
<div>
  <button class="btn no-animation text-base h-auto min-h-fit px-2 py-2 btn-primary focus:bg-primary-focus focus:border-primary-focus hover:bg-primary-focus disabled:opacity-40 disabled:bg-primary disabled:text-white">
    Open demo
  </button>
  <dialog class="modal items-end sm:items-center"
          id="modal-demo-basic"
  >
    <div class="modal-box flex flex-col rounded-md p-8 pt-6 w-full max-w-none sm:max-w-xl overflow-hidden">
      <form method="dialog"
            class="flex justify-end"
      >
        <button class="btn btn-xs btn-square btn-ghost"
                aria-label="close modal"
        >
          <svg xmlns="http://www.w3.org/2000/svg"
               viewbox="0 0 24 24"
               fill="none"
               stroke="currentColor"
               class="inline-block  stroke-foreground icon-cross"
               width="16"
               aria-hidden="true"
          >
            <path d="M18,6h0a1,1,0,0,0-1.414,0L12,10.586,7.414,6A1,1,0,0,0,6,6H6A1,1,0,0,0,6,7.414L10.586,12,6,16.586A1,1,0,0,0,6,18H6a1,1,0,0,0,1.414,0L12,13.414,16.586,18A1,1,0,0,0,18,18h0a1,1,0,0,0,0-1.414L13.414,12,18,7.414A1,1,0,0,0,18,6Z"
                  stroke="none"
            >
            </path>
          </svg>
        </button>
      </form>
      <div class="mt-4 overflow-y-auto">
        <h2 class="font-semibold text-2xl mb-8">
          Modal title
        </h2>
        <p>
          Press ESC key or click outside to close
        </p>
      </div>
    </div>
    <form method="dialog"
          class="modal-backdrop"
          aria-hidden="true"
    >
      <button class="cursor-default"
              tabindex="-1"
      >
      </button>
    </form>
  </dialog>
</div>
`;
